pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -20,56 +21,112 @@
 #include <sys/stat.h>
 #include <sys/wait.h>
 
-char *env_whitelist[] = {"TERM",
-                         "DISPLAY",
-                         "XAUTHORITY",
-                         "HOME",
-                         "XDG_RUNTIME_DIR",
-                         "LANG",
-                         "SSL_CERT_FILE",
-                         "DBUS_SESSION_BUS_ADDRESS"};
-
-char **env_build(char *names[], size_t len) {
-  char *env, **ret = malloc((len + 1) * sizeof(char *)), **ptr = ret;
-
-  for (size_t i = 0; i < len; i++) {
-    if ((env = getenv(names[i]))) {
-      if (asprintf(ptr++, "%s=%s", names[i], env) < 0)
-        errorf(EX_OSERR, "asprintf");
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+char *env_blacklist[] = {};
+
+char **env_filter(char *envp[]) {
+  char **filtered_envp = malloc(sizeof(*envp));
+  size_t n = 0;
+
+  while (*envp != NULL) {
+    bool blacklisted = false;
+
+    for (size_t i = 0; i < LEN(env_blacklist); i++) {
+      if (!strncmp(*envp, env_blacklist[i], strlen(env_blacklist[i]))) {
+        blacklisted = true;
+        break;
+      }
+    }
+
+    if (!blacklisted) {
+      filtered_envp = realloc(filtered_envp, (n + 2) * sizeof(*envp));
+
+      if (filtered_envp == NULL)
+        errorf(EX_OSERR, "realloc");
+
+      filtered_envp[n++] = *envp;
     }
+
+    envp++;
+  }
+
+  filtered_envp[n] = NULL;
+  return filtered_envp;
+}
+
+void bind(char *from, char *to) {
+  if (mkdir(to, 0755) < 0)
+    errorf(EX_IOERR, "mkdir");
+
+  if (mount(from, to, "bind", MS_BIND | MS_REC, NULL) < 0)
+    errorf(EX_OSERR, "mount");
+}
+
+char *strjoin(char *dir, char *name) {
```

Comment:
This whole function could just be `asprintf("%s%s", dir, name)`

or in C++: `dir + name`. ;)

Reply:
I know it could be `asprintf`, but it's cleaner this way. Check previous chrootenv version, I've used `asprintf`.

> or in C++: [...]

You won't succumb me to the dark side! Heh.

Reply:
How is it cleaner to write 13 lines of code to replace 1 line?

Reply:
It has better semantics. There is no template, only concatenation.

Reply:
But every C programmer who reads `asprintf(&ret, "%s%s", x, y)` immediately knows it's just concatenating two strings.

And in fact you're using this in precisely one place, with a constant string as the other arg. Just
````
char *host_target = strjoin("host/", prefix_dirent->d_name)
````
->
````
char *host_target;
asprintf(&host_target, "host/%s", prefix_dirent->d_name);
````
to make it even more simpler.

Reply:
OK, point taken. `errorf` uses the same DSL anyway so it seems there is not much use in implementing simpler `asprintf`.